docs/architecture/adr/0025-ts-deprecate-enums.md

```diff
@@ -0,0 +1,196 @@
+  lint.
+
+## Decision Outcome
+
+Chosen option: **Deprecate enum use**
+
+### Positive Consequences
+
+- Allows for cases where autogenerated code introduces an enum by necessity.
+- Developers receive a warning in their IDE to discourage new enums.
+- The warning can direct them to our contributing docs, where they can learn typesafe alternatives.
+- Our compiled code size decreases when enums are replaced.
+- If all teams eliminate enums in practice, the warning can be increased to an error.
+
+### Negative Consequences
+
+- Unnecessary usage may persist indefinitely on teams carrying a high tech debt.
+- The lint increased the number of FIXME comments in the code by about 10%.
+
+### Plan
+
+- Update contributing docs with patterns and best practices for enum replacement.
+- Update the reporting level of the lint to "warning".
+
+[computed-property-names]:
+  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#computed_property_names
+[literal-type]: https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#literal-types
+[no-enum-lint]: https://github.com/bitwarden/clients/blob/main/libs/eslint/platform/no-enums.mjs
+[no-member-fields-fixed]:
+  https://devblogs.microsoft.com/typescript/announcing-typescript-5-8-beta/#preserved-computed-property-names-in-declaration-files
+
+## Appendix A: Mapped Types and Enum-likes
+
+Mapped types cannot determine that a mapped enum-like object is fully assigned. Code like the
+following causes a compiler error:
+
+```ts
+const instance: Record<CipherType, boolean> = {
+  [CipherType.Login]: true,
+  [CipherType.SecureNote]: false,
+  [CipherType.Card]: true,
+  [CipherType.Identity]: true,
+  [CipherType.SshKey]: true,
+};
+```
+
+#### Why does this happen?
+
+The members of `const _CipherType` all have a [literal type][literal-type]. `_CipherType.Login`, for
+example, has a literal type of `1`. `type CipherType` maps over these members, aggregating them into
+the structural type `1 | 2 | 3 | 4 | 5`.
+
+`const CipherType` asserts its members have `type CipherType`, which overrides the literal types the
+compiler inferred for the member in `const _CipherType`. The compiler sees the type of
+`CipherType.Login` as `type CipherType` (which aliases `1 | 2 | 3 | 4 | 5`).
+
+Now consider a mapped type definition:
+
+```ts
+// `MappedType` is structurally identical to Record<CipherType, boolean>
+type MappedType = { [K in CipherType]: boolean };
+```
+
+When the compiler examines `instance`, it only knows that the type of each of its members is
+`CipherType`. That is, the type of `instance` to the compiler is
+`{ [K in 1 | 2 | 3 | 4 | 5]?: boolean }`. This doesn't sufficiently overlap with `MappedType`, which
+is looking for `{ [1]: boolean, [2]: boolean, [3]: boolean, [4]: boolean, [5]: boolean }`. The
+failure occurs, because the inferred type can have fewer fields than `MappedType`.
+
+### Workarounds
```

Comment:
thought: I feel like none of these workarounds actually fix the issue which causes less than possible static type-safety. Are we doing this because we value consistency more than the possibility of statically checked mapped types? Should we consider an "escape hatch" that allows access to the literally typed member in our recommended approach?

```ts
const instance: Record<CipherType, boolean> = {
  [CipherType[Literal].Login]: true,
  [CipherType[Literal].SecureNote]: false,
  [CipherType[Literal].Card]: true,
  [CipherType[Literal].Identity]: true,
  [CipherType[Literal].SshKey]: true,
};
```

I guess this is a perfect topic for the arch meeting :)